modelforge/potential/neighbors.py

```diff
@@ -0,0 +1,418 @@
+        if self.builds == 0:
+            self.box_vectors = data.box_vectors
+
+        box_changed = torch.any(self.box_vectors != data.box_vectors)
+
+        # avoid reinitializing indices if they are already set and haven't changed
+        if self.indices.shape[0] != n:
+            self.box_vectors = data.box_vectors
+            self.positions_old = positions
+            self._init_pairs(n, positions.device)
+            r_ij, d_ij = self._build_nlist(positions, data.box_vectors)
+        elif box_changed:
+            # if the box vectors have changed, we need to rebuild the nlist
+            # but do not need to regenerate the pairs
+            self.box_vectors = data.box_vectors
+            self.positions_old = positions
+            r_ij, d_ij = self._build_nlist(positions, data.box_vectors)
+        elif self._check_nlist(positions, data.box_vectors):
+            self.positions_old = positions
+            r_ij, d_ij = self._build_nlist(positions, data.box_vectors)
+        else:
+            r_ij, d_ij = self.displacement_function(
+                positions[self.nlist_pairs[0]],
+                positions[self.nlist_pairs[1]],
+                data.box_vectors,
+            )
+
+        in_cutoff = (d_ij <= self.cutoff).squeeze()
+        total_pairs = in_cutoff.sum()
+
+        if self.only_unique_pairs:
+            # using this instead of torch.stack to ensure that if we only have a single pair
+            # we don't run into an issue with shapes.
+
+            pairs = torch.zeros(
+                2, total_pairs, dtype=torch.int64, device=positions.device
+            )
+
+            pairs[0] = self.nlist_pairs[0][in_cutoff]
+            pairs[1] = self.nlist_pairs[1][in_cutoff]
+
+            return PairlistData(
+                pair_indices=pairs,
+                d_ij=d_ij[in_cutoff],
+                r_ij=r_ij[in_cutoff],
+            )
+
+        else:
+
+            r_ij_full = torch.zeros(
+                total_pairs * 2, 3, dtype=positions.dtype, device=positions.device
+            )
+            d_ij_full = torch.zeros(
+                total_pairs * 2, 1, dtype=positions.dtype, device=positions.device
+            )
+
+            temp = r_ij[in_cutoff]
+            r_ij_full[0:total_pairs] = temp
+            r_ij_full[total_pairs : 2 * total_pairs] = -temp
+
+            del r_ij
+
+            temp = d_ij[in_cutoff]
+            d_ij_full[0:total_pairs] = temp
+            d_ij_full[total_pairs : 2 * total_pairs] = temp
+
+            del d_ij
+
+            temp1 = self.nlist_pairs[0][in_cutoff]
+            temp2 = self.nlist_pairs[1][in_cutoff]
+
+            pairs = torch.zeros(
+                2, total_pairs * 2, dtype=torch.int64, device=positions.device
+            )
+
+            pairs[0][0:total_pairs] = temp1
+            pairs[1][0:total_pairs] = temp2
+            pairs[0][total_pairs : 2 * total_pairs] = temp2
+            pairs[1][total_pairs : 2 * total_pairs] = temp1
+
```

Comment:
can we extract this logic in a function?

and as I mentioned above, I think the special case should be `unique_pairs`, not non-unique pairs. IMO, when we use the meshgrid, non-unqie pairs are simply the flattened `grid_x` and `grid_y` outputs of the meshgrid, where the diagonal is removed.

Reply:
The reason why I have this only calculate for unique pairs, and then copy if we need non-unique pairs is efficiency.  The copy operation is negligible cost compare to doing twice as many computations (pretty much exactly a factor of 2).   It would probably be good to make this a separate function since it is used in both the brute and vert list.

Reply:
This is now in a separate function in each class to make it a little cleaner.  However, the same logic still applies of copying unique pairs data into non-unique pair data.

Reply:
oh, thank you for clarifying! Please ignore my post above in that case! That is a sufficient reason to be slightly non-intuitive :-)